src/vm/methodtable.cpp

```diff
@@ -9850,7 +9850,7 @@ MethodTable::TryResolveConstraintMethodApprox(
                 pMD = pCanonMT->GetMethodDescForInterfaceMethod(thPotentialInterfaceType, pGenInterfaceMD, FALSE /* throwOnConflict */);
 
                 // See code:#TryResolveConstraintMethodApprox_DoNotReturnParentMethod
-                if ((pMD != NULL) && !pMD->GetMethodTable()->IsValueType())
+                if ((pMD != NULL) && !pMD->GetMethodTable()->IsValueType() && !pMD->IsInterface())
```

Comment:
What happens if the method that we resolve to hear is a generic, instantiated over canon. Does this still work here?

Reply:
I've added test coverage for the generic methods case, but I'm not sure I understood the concern - we strip the method instantiation before entering this code path, so the particular instantiation of the generic method shouldn't matter.

Reply:
I meant type generic. You know, the notorious case that looks like

```csharp

interface ITest<A>
{
 string M() { return "default implementation"; }
}


interface IDerivedTest<A> : ITest<A>
{
    override ITest<A>.M() { return "derived default implementation"; }
}

struct Test<X, Y> : ITest<X> : IDerivedTest<Y>
{
}

...
string SomeGenericMethod<W, Z> where W : ITest<Z>
{
    return W.M();
}

...
SomeGenericMethod<Test<object,string>, object>() -> should yield "default implementation"
and 
SomeGenericMethod<Test<object,object>, object>() -> should yield "derived default implementation"
```

Now that I think about it though, I think you've discussed already disabling this weird situation in any case, so it shouldn't matter.

Reply:
Yes, this case is going to throw because it requires a "boxing stub". `constrained2.il` has a test for a similar situation (we're in a situation where the dispatch is ambiguous at compile time between an implementation provided by the value type and the default implementation), but at runtime we can call the implementation on the value type (there's also a try/catch for the "boxing stub" case that works both if the runtime does the right thing, or when it throws).